src/matrices.jl

```diff
@@ -0,0 +1,53 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENCE in the project root.
+# ------------------------------------------------------------------
+
+abstract type CoDaMatrix{T} end
+
+"""
+    JMatrix{T}
+
+Square matrix of ones.
+"""
+struct JMatrix{T} <: CoDaMatrix{T} end
+JMatrix{T}(d::Integer) where {T} = Ones{T}(d, d)
+(J::JMatrix{T})(d::Integer) where {T} = JMatrix{T}(d)
+const J = JMatrix{Bool}()
+(*)(::JMatrix, v::AbstractVector) = Fill(sum(v), length(v))
+(*)(v::Adjoint{<:Any, <:AbstractVector}, J::JMatrix) = (J*v')'
```

Comment:
Try to replace `Adjoint` by `AbstractMatrix`:

```
supertype(Adjoint) == supertype(Transpose) == AbstractMatrix
```

that way we can left-multiply by any "row vector". I would also try to add a method for `(::JMatrix, ::AbstractMatrix)` besides the one you wrote for `(::JMatrix, ::AbstractVector)`.

Reply:
I'll try to make them work like in the first commit of the PR, then we can benchmark the difference.

Reply:
`Fill`s get evaluated to an `Array` when composed as a matrix:

```julia> [Fill(1, 10, 1) Fill(2, 10, 1) Fill(3, 10, 1)]
10×3 Matrix{Int64}:
 1  2  3
 1  2  3
 ⋮
 1  2  3

julia> typeof([Fill(1, 10, 1) Fill(2, 10, 1) Fill(3, 10, 1)])
Matrix{Int64} (alias for Array{Int64, 2})
```

Due to possible unwanted allocations, this means that we might be better off distributing the multiplication and  adding the resulting matrices instead of using an appending-based approach. 

Further testing is required to find out the fastest methods.

Reply:
We can use BlockArrays.jl later if that shows to be an issue. Let's finalize this round of review and then we start a fresh PR just for optimizations 👍🏽